dev/src/expression.ts

```diff
@@ -0,0 +1,1425 @@
+  }
+}
+
+export class Function extends Expr {
+  exprType: ExprType = 'Function';
+  constructor(
+    private name: string,
+    private params: Expr[]
+  ) {
+    super();
+  }
+
+  _toProto(serializer: Serializer): api.IValue {
+    return {
+      functionValue: {
+        name: this.name,
+        args: this.params.map(p => serializer.encodeValue(p)!),
+      },
+    };
+  }
+}
+
+export class Add extends Function {
+  constructor(
+    private left: Expr,
+    private right: Expr
+  ) {
+    super('add', [left, right]);
+  }
+}
+
+export class Subtract extends Function {
+  constructor(
+    private left: Expr,
+    private right: Expr
+  ) {
+    super('subtract', [left, right]);
+  }
+}
+
+export class Multiply extends Function {
+  constructor(
+    private left: Expr,
+    private right: Expr
+  ) {
+    super('multiply', [left, right]);
+  }
+}
+
+export class Divide extends Function {
+  constructor(
+    private left: Expr,
+    private right: Expr
+  ) {
+    super('divide', [left, right]);
+  }
+}
+
+export class Eq extends Function implements FilterCondition {
+  constructor(
+    private left: Expr,
+    private right: Expr
+  ) {
+    super('eq', [left, right]);
+  }
+  filterable = true as const;
+}
+
+class Neq extends Function implements FilterCondition {
+  constructor(
+    private left: Expr,
+    private right: Expr
+  ) {
+    super('neq', [left, right]);
+  }
+  filterable = true as const;
+}
+
+class Lt extends Function implements FilterCondition {
+  constructor(
+    private left: Expr,
+    private right: Expr
+  ) {
+    super('lt', [left, right]);
+  }
+  filterable = true as const;
+}
+
+class Lte extends Function implements FilterCondition {
+  constructor(
+    private left: Expr,
+    private right: Expr
+  ) {
+    super('lte', [left, right]);
+  }
+  filterable = true as const;
+}
+
+class Gt extends Function implements FilterCondition {
+  constructor(
+    private left: Expr,
+    private right: Expr
+  ) {
+    super('gt', [left, right]);
+  }
+  filterable = true as const;
+}
+
+class Gte extends Function implements FilterCondition {
+  constructor(
+    private left: Expr,
+    private right: Expr
+  ) {
+    super('gte', [left, right]);
+  }
+  filterable = true as const;
+}
+
+class ArrayConcat extends Function {
+  constructor(
+    private array: Expr,
+    private elements: Expr[]
+  ) {
+    super('array_concat', [array, ...elements]);
+  }
+}
+
+class ArrayContains extends Function implements FilterCondition {
+  constructor(
+    private array: Expr,
+    private element: Expr
+  ) {
+    super('array_contains', [array, element]);
+  }
+  filterable = true as const;
+}
+
+class ArrayContainsAll extends Function implements FilterCondition {
+  constructor(
+    private array: Expr,
+    private values: Expr[]
+  ) {
+    super('array_contains_all', [array, new ListOfExprs(values)]);
+  }
+  filterable = true as const;
+}
+
+class ArrayContainsAny extends Function implements FilterCondition {
```

Comment:
Is FilterCondition always a function that returns a boolean? If so, would it be useful to make a more generic BooleanFunction interface in case there is a future use case for boolean functions outside of filtering?